examples/python/learning_pytorch.py

```diff
@@ -48,230 +47,250 @@
+            best_action_index = agent.get_action(state)
+
+            game.make_action(actions[best_action_index], frame_repeat)
+        r = game.get_total_reward()
+        test_scores.append(r)
+
+    test_scores = np.array(test_scores)
+    print("Results: mean: %.1f +/- %.1f," % (
+        test_scores.mean(), test_scores.std()), "min: %.1f" % test_scores.min(),
+          "max: %.1f" % test_scores.max())
+
+
+def run(game, agent, actions, num_epochs, frame_repeat, steps_per_epoch=2000):
+    """
+    Run num epochs of training episodes.
+    Skip frame_repeat number of frames after each action.
+    """
+    scores = []
+    start_time = time()
+
+    for epoch in range(num_epochs):
+        game.new_episode()
+        train_scores = []
+        global_step = 0
+        print("\nEpisode #" + str(epoch + 1))
+
+        for _ in trange(steps_per_epoch, leave=False):
+            state = preprocess(game.get_state().screen_buffer)
+            action = agent.get_action(state)
+            reward = game.make_action(actions[action], frame_repeat)
+            done = game.is_episode_finished()
+
+            if not done:
+                next_state = preprocess(game.get_state().screen_buffer)
+            else:
+                next_state = np.zeros((1, 30, 45)).astype(np.float32)
+
+            agent.append_memory(state, action, reward, next_state, done)
+
+            if global_step > agent.batch_size:
+                agent.train()
+
+            if done:
+                scores.append(game.get_total_reward())
+                train_scores.append(game.get_total_reward())
+                game.new_episode()
+
+            global_step += 1
 
-        self.capacity = capacity
-        self.size = 0
-        self.pos = 0
+        agent.update_target_net()
+        train_scores = np.array(train_scores)
 
-    def add_transition(self, s1, action, s2, isterminal, reward):
-        self.s1[self.pos, 0, :, :] = s1
-        self.a[self.pos] = action
-        if not isterminal:
-            self.s2[self.pos, 0, :, :] = s2
-        self.isterminal[self.pos] = isterminal
-        self.r[self.pos] = reward
+        print("Results: mean: %.1f +/- %.1f," % (train_scores.mean(), train_scores.std()),
+              "min: %.1f," % train_scores.min(), "max: %.1f," % train_scores.max())
 
-        self.pos = (self.pos + 1) % self.capacity
-        self.size = min(self.size + 1, self.capacity)
+        test(game, agent)
+        if save_model:
+            print("Saving the network weigths to:", model_savefile)
+            torch.save(agent.q_net, model_savefile)
+        print("Total elapsed time: %.2f minutes" % ((time() - start_time) / 60.0))
 
-    def get_sample(self, sample_size):
-        i = sample(range(0, self.size), sample_size)
-        return self.s1[i], self.a[i], self.s2[i], self.isterminal[i], self.r[i]
+    game.close()
+    return agent, game
 
 
-class Net(nn.Module):
+class DuelQNet(nn.Module):
```

Comment:
A docstring here clarifying that this is Dueling DQN would be nice (and link to the paper).

Reply:
Done